### PR TITLE
Add python wrapper for  CTC greedy decoder and edit distance evaluator

### LINKED ISSUE #7596: Add greedy CTC evaluator python API
This issue depend on https://github.com/PaddlePaddle/Paddle/pull/7527
CTC evaluator = top-k_op + ctc_align_op + edit_distance_op

Test script:
```
import numpy as np
import paddle.v2 as paddle
import paddle.v2.fluid as fluid
from paddle.v2.fluid import core

x = fluid.layers.data(name='x', shape=[8], dtype='float32')
y = fluid.layers.data(name='y', shape=[1], dtype='float32')
ctc_result = fluid.layers.ctc_greedy_decoder(input=x, blank=0)
edit_distance = fluid.evaluator.EditDistance(input=ctc_result,label=y)
print "step1"

place = fluid.CPUPlace()
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())
print "step2"
edit_distance.reset(exe)
batch_num = 2
for i in range(batch_num):
    print "step3"
    y_data = np.random.randint(0, 8, [7, 1])
    y_lod = [[0, 2, 4, 7]]
    y_tensor = core.LoDTensor()
    y_tensor.set(y_data, place)
    y_tensor.set_lod(y_lod)

    x_data = np.random.uniform(0.1, 1, [11, 8]).astype("float32")
    x_lod = [[0, 3, 5, 11]]
    x_tensor = core.LoDTensor()
    x_tensor.set(x_data, place)
    x_tensor.set_lod(x_lod)

    cost, = exe.run(fluid.default_main_program(),
                              feed={
                                   'x': x_tensor,
                                   'y': y_tensor
                                },
                                fetch_list=edit_distance.metrics)
    pass_error = edit_distance.eval(exe)
    print "cost: %s" % cost
    print "pass_id=%d; pass_error=%s" % (i, str(pass_error))

pass_error = edit_distance.eval(exe)
print "total_pass_error=%s" % str(pass_error)
```

### PR DESCRIPTION
**Pull request by wanghaoshuang**
fix #7596

### CODE REVIEW

**Comment by kuke** [comment]
I suggest to divide this operator into two, ctc_greedy_decoder and edit_distance_evaluator. Please always remember that there is another decoding method which is beam search. Users may want to output the decoding result directly, or they want to use beam search decoding and then evaluate the result. It would be unfeasible if wrapping greedy decoding and error evaluating in one operator.

**Comment by wanghaoshuang** [comment]
@kuke Thanks for your reminder.

File: python/paddle/v2/fluid/evaluator.py [hunk 1]
```diff
@@ -204,3 +204,37 @@ def eval(self, executor, eval_program=None):
...
+        if main_program.current_block().idx != 0:
+            raise ValueError("You can only invoke Evaluator in root block")
+
+        self.total_error = self.create_state(
+            dtype='float32', shape=[1], suffix='total')
+        self.batch_num = self.create_state(
+            dtype='float32', shape=[1], suffix='total')
+        error = layers.edit_distance(input=input, label=label)
+        error = layers.cast(x=error, dtype='float32')
+        mean_error = layers.mean(x=error)
```
Comment (qingqing01) [inline python/paddle/v2/fluid/evaluator.py]: Better to make consistency with `Accuracy` evaluator, do not calculate the average mean error of current mini-batch.  Just accumulate the batch size and mini-batch error here.

File: python/paddle/v2/fluid/evaluator.py [hunk 2]
```diff
@@ -204,3 +204,37 @@ def eval(self, executor, eval_program=None):
             [precision], dtype='float32'), np.array(
                 [recall], dtype='float32'), np.array(
                     [f1_score], dtype='float32')
+
+
+class EditDistance(Evaluator):
+    """
+    Average edit distance error for multiple mini-batches.
```
Comment (qingqing01) [inline python/paddle/v2/fluid/evaluator.py]: Need more comments, How to usage, and what the returned value by `eval` means.

File: python/paddle/v2/fluid/layers/nn.py [hunk 3]
```diff
@@ -1863,6 +1864,140 @@ def matmul(x, y, transpose_x=False, transpose_y=False, name=None):
...
+        input(Variable): The indices for hypothesis strings.
+
+        label(Variable): The indices for reference strings.
+
+        normalized(bool): Indicated whether to normalize the edit distance by the length of reference string.
+
+        tokens(list): Tokens that should be removed before calculating edit distance.
+
+    Returns:
+        Variable: sequence-to-sequence edit distance loss in shape [batch_size, 1].
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: Remove loss.

File: python/paddle/v2/fluid/layers/nn.py [hunk 4]
```diff
@@ -1863,6 +1864,140 @@ def matmul(x, y, transpose_x=False, transpose_y=False, name=None):
     return out
 
 
+def edit_distance(input, label, normalized=False, tokens=None, name=None):
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: tokens -> ignored_tokens ?

File: python/paddle/v2/fluid/layers/nn.py [hunk 5]
```diff
@@ -1863,6 +1864,140 @@ def matmul(x, y, transpose_x=False, transpose_y=False, name=None):
...
+
+    Args:
+
+        input(Variable): The indices for hypothesis strings.
+
+        label(Variable): The indices for reference strings.
+
+        normalized(bool): Indicated whether to normalize the edit distance by the length of reference string.
+
+        tokens(list): Tokens that should be removed before calculating edit distance.
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: tokens(list) -> list of int

File: python/paddle/v2/fluid/evaluator.py (hunk 1 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/evaluator.py]: Fixed.

File: python/paddle/v2/fluid/layers/nn.py (hunk 4 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/layers/nn.py]: Thx. Fixed.

File: python/paddle/v2/fluid/layers/nn.py (hunk 5 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/layers/nn.py]: Fixed.

File: python/paddle/v2/fluid/layers/nn.py (hunk 3 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/layers/nn.py]: Thx. Done.

File: python/paddle/v2/fluid/evaluator.py (hunk 2 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/evaluator.py]: Thk. Done.

**Pull request merged**